### PR TITLE
[red-knot] fix: when simplifying union, True & False -> instance(bool)

### LINKED ISSUE #13642: [red-knot] simplification of 'True | False' is bool class (instead of bool instance)
Extracted [this comment](https://github.com/astral-sh/ruff/pull/13615#discussion_r1788554559) from @AlexWaygood to keep track of this bug

> Ah. @carljm, I think this might actually reveal a bug in main. The union of Literal[True] | Literal[False] should surely simplify to Instance("bool") rather than Class("bool"). If an object has type Literal[True], it indicates that it the object is an instance of bool, not that it is the bool class itself.

### PR DESCRIPTION
**Pull request by Slyces**
## Summary

Fixes #13642

## Test Plan

- Edit the existing test
- Add a test in `infer.rs` using `reveal_type`

### CODE REVIEW

**Review by AlexWaygood — COMMENTED** [review]
Looks great! We could also add a test to `infer.rs` using reveal type, with a snippet like this:

```py
from typing_extensions import reveal_type

def returns_bool() -> bool: ...

if returns_bool():
    x = True
else:
    x = False

reveal_type(x)
```

Calling `assert_file_diagnostics` on a file with this example in it should give us ```["Revealed type is `bool`"]``` rather than ```["Revealed type is `Literal[bool]`"]```.

**Comment by Slyces** [comment]
Great comment, I was thinking that indeed the current tests are tightly coupled. Doing this now :)

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 1]
```diff
@@ -4548,6 +4548,31 @@ mod tests {
...
+    #[test]
+    fn simplify_true_and_false_to_bool() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            from typing_extensions import reveal_type
+
+            def returns_bool() -> bool: ...
```
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types/infer.rs:4560]: Ah, at some point we'll start checking return types, which might lead to us emitting a spurious diagnostic in this test snippet. We can make it robust to future improvements like so:

```suggestion
            def returns_bool() -> bool:
                return True
```

**Review by AlexWaygood — APPROVED** [review]
Thanks!

**Comment by carljm** [comment]
Thank you!!

**Pull request merged**